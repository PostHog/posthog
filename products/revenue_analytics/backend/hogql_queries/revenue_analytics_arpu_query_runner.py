from decimal import Decimal

from posthog.hogql import ast
from posthog.hogql.query import execute_hogql_query
from posthog.schema import (
    HogQLQueryResponse,
    CachedRevenueAnalyticsArpuQueryResponse,
    RevenueAnalyticsArpuQueryResponse,
    RevenueAnalyticsArpuQuery,
)
from posthog.models.filters.mixins.utils import cached_property
from posthog.hogql_queries.utils.timestamp_utils import format_label_date

from .revenue_analytics_query_runner import RevenueAnalyticsQueryRunner
from products.revenue_analytics.backend.views import RevenueAnalyticsInvoiceItemView


class RevenueAnalyticsArpuQueryRunner(RevenueAnalyticsQueryRunner):
    query: RevenueAnalyticsArpuQuery
    response: RevenueAnalyticsArpuQueryResponse
    cached_response: CachedRevenueAnalyticsArpuQueryResponse

    def to_query(self) -> ast.SelectQuery:
        with self.timings.measure("revenue_subquery"):
            revenue_subquery = self._revenue_subquery()
            if revenue_subquery is None:
                return ast.SelectQuery.empty(columns=["breakdown_by", "period_start", "value"])

        with self.timings.measure("customer_count_subquery"):
            customer_count_subquery = self._customer_count_subquery()
            if customer_count_subquery is None:
                return ast.SelectQuery.empty(columns=["breakdown_by", "period_start", "value"])

        return ast.SelectQuery(
            select=[
                ast.Alias(alias="breakdown_by", expr=ast.Field(chain=["revenue", "breakdown_by"])),
                ast.Alias(alias="period_start", expr=ast.Field(chain=["revenue", "period_start"])),
                ast.Alias(
                    alias="value",
                    expr=ast.Call(
                        name="if",
                        args=[
                            ast.Or(
                                exprs=[
                                    ast.Call(name="isNull", args=[ast.Field(chain=["customer_count", "count"])]),
                                    ast.CompareOperation(
                                        op=ast.CompareOperationOp.Eq,
                                        left=ast.Field(chain=["customer_count", "count"]),
                                        right=ast.Constant(value=0),
                                    ),
                                ],
                            ),
                            ast.Constant(value=0),
                            ast.Call(
                                name="divide",
                                args=[
                                    ast.Field(chain=["revenue", "amount"]),
                                    ast.Field(chain=["customer_count", "count"]),
                                ],
                            ),
                        ],
                    ),
                ),
            ],
            select_from=ast.JoinExpr(
                alias="revenue",
                table=revenue_subquery,
                next_join=ast.JoinExpr(
                    alias="customer_count",
                    table=customer_count_subquery,
                    join_type="LEFT JOIN",
                    constraint=ast.JoinConstraint(
                        constraint_type="ON",
                        expr=ast.And(
                            exprs=[
                                ast.CompareOperation(
                                    op=ast.CompareOperationOp.Eq,
                                    left=ast.Field(chain=["revenue", "period_start"]),
                                    right=ast.Field(chain=["customer_count", "period_start"]),
                                ),
                                ast.CompareOperation(
                                    op=ast.CompareOperationOp.Eq,
                                    left=ast.Field(chain=["revenue", "breakdown_by"]),
                                    right=ast.Field(chain=["customer_count", "breakdown_by"]),
                                ),
                            ],
                        ),
                    ),
                ),
            ),
            # Return sorted by period_start/day_start, and then for each individual day we put the maximum first (value)
            # This will allow us to return the list sorted according to the numbers in the first day
            # Finally sort by breakdown_by for the rare cases where they tie (usually at 0 revenue)
            order_by=[
                ast.OrderExpr(expr=ast.Field(chain=["period_start"]), order="ASC"),
                ast.OrderExpr(expr=ast.Field(chain=["value"]), order="DESC"),
                ast.OrderExpr(expr=ast.Field(chain=["breakdown_by"]), order="ASC"),
            ],
            # Need a huge limit because we need (dates x breakdown)-many rows to be returned
            limit=ast.Constant(value=10000),
        )

    # NOTE: This is simply revenue in the current period and not the MRR
    # We might wanna let customers choose between MRR and revenue in the future
    def _revenue_subquery(self) -> ast.SelectQuery | None:
        if self.revenue_subqueries.invoice_item is None:
            return None

        with self.timings.measure("dates_expr"):
            dates_expr = self._dates_expr()

        query = ast.SelectQuery(
            select=[
                ast.Alias(
                    alias="breakdown_by",
                    expr=ast.Field(chain=[RevenueAnalyticsInvoiceItemView.get_generic_view_alias(), "source_label"]),
                ),
                ast.Alias(alias="period_start", expr=dates_expr),
                ast.Alias(alias="amount", expr=ast.Call(name="sum", args=[ast.Field(chain=["amount"])])),
            ],
            select_from=self._append_joins(
                ast.JoinExpr(
                    alias=RevenueAnalyticsInvoiceItemView.get_generic_view_alias(),
                    table=self.revenue_subqueries.invoice_item,
                ),
                self.joins_for_properties(RevenueAnalyticsInvoiceItemView),
            ),
            where=ast.And(
                exprs=[
                    self.timestamp_where_clause(
                        chain=[RevenueAnalyticsInvoiceItemView.get_generic_view_alias(), "timestamp"]
                    ),
                    *self.where_property_exprs,
                ]
            ),
            group_by=[ast.Field(chain=["breakdown_by"]), ast.Field(chain=["period_start"])],
        )

        # Limit to 2 group bys at most for performance reasons
        # This is also implemented in the frontend, but let's guarantee it here too
        with self.timings.measure("append_group_by"):
            for group_by in self.query.groupBy[:2]:
                query = self._append_group_by(query, RevenueAnalyticsInvoiceItemView, group_by)

        return query

    def _customer_count_subquery(self) -> ast.SelectQuery | None:
        if self.revenue_subqueries.customer is None:
            return None

        with self.timings.measure("dates_expr"):
            dates_expr = self._dates_expr()

        query = ast.SelectQuery(
            select=[
                ast.Alias(
                    alias="breakdown_by",
                    expr=ast.Field(chain=[RevenueAnalyticsInvoiceItemView.get_generic_view_alias(), "source_label"]),
                ),
                ast.Alias(alias="period_start", expr=dates_expr),
                ast.Alias(
                    alias="count",
                    expr=ast.Call(
                        name="countIf",
                        distinct=True,
                        args=[
                            ast.Field(chain=[RevenueAnalyticsInvoiceItemView.get_generic_view_alias(), "customer_id"]),
                            ast.CompareOperation(
                                op=ast.CompareOperationOp.Eq,
                                left=ast.Call(
                                    name=f"toStartOf{self.query_date_range.interval_name.title()}",
                                    args=[
                                        ast.Field(
                                            chain=[
                                                RevenueAnalyticsInvoiceItemView.get_generic_view_alias(),
                                                "timestamp",
                                            ]
                                        ),
                                    ],
                                ),
                                right=ast.Field(chain=["period_start"]),
                            ),
                        ],
                    ),
                ),
            ],
            select_from=self._append_joins(
                ast.JoinExpr(
                    alias=RevenueAnalyticsInvoiceItemView.get_generic_view_alias(),
                    table=self.revenue_subqueries.invoice_item,
                ),
                self.joins_for_properties(RevenueAnalyticsInvoiceItemView),
            ),
            group_by=[
                ast.Field(chain=["breakdown_by"]),
                ast.Field(chain=["period_start"]),
            ],
            where=ast.And(
                exprs=[
                    self.timestamp_where_clause(
                        [RevenueAnalyticsInvoiceItemView.get_generic_view_alias(), "timestamp"]
                    ),
                    *self._parsed_where_property_exprs,
                ]
            ),
        )

        # Limit to 2 group bys at most for performance reasons
        # This is also implemented in the frontend, but let's guarantee it here too
        with self.timings.measure("append_group_by"):
            for group_by in self.query.groupBy[:2]:
                query = self._append_group_by(query, RevenueAnalyticsInvoiceItemView, group_by)

        return query

    @cached_property
    def _parsed_where_property_exprs(self) -> list[ast.Expr]:
        where_property_exprs = self.where_property_exprs

        # We can't join from a subscription back to the invoice item table,
        # so let's remove any where property exprs that are comparing to the invoice item table
        return [
            expr
            for expr in where_property_exprs
            if isinstance(expr, ast.CompareOperation)
            and isinstance(expr.left, ast.Field)
            and expr.left.chain[0] != RevenueAnalyticsInvoiceItemView.get_generic_view_alias()
        ]

    def _build_results(self, response: HogQLQueryResponse) -> list[dict]:
        # We want the result to look just like the Insights query results look like to simplify our UI
        # First, let's generate all of the dates/labels because they'll be exactly the same for all of the results
        all_dates = self.query_date_range.all_values()
        days = [date.strftime("%Y-%m-%d") for date in all_dates]
        labels = [format_label_date(item, self.query_date_range, self.team.week_start_day) for item in all_dates]

        # We can also group the results we have by a tuple of (breakdown_by, period_start)
        # This will allow us to easily query the results by breakdown_by and period_start
        # and then we can just add the data to the results
        # [0, 1, 2] -> [value, period_start, breakdown_by]
        grouped_results: dict[tuple[str, str], Decimal] = {}
        breakdowns = []
        for breakdown_by, period_start, count in response.results:
            if breakdown_by not in breakdowns:
                breakdowns.append(breakdown_by)

            grouped_results[(breakdown_by, period_start.strftime("%Y-%m-%d"))] = count

        return [
            {
                "action": {
                    "days": all_dates,
                    "id": breakdown,
                    "name": breakdown,
                },
                "data": [grouped_results.get((breakdown, day), 0) for day in days],
                "days": days,
                "label": breakdown,
                "labels": labels,
            }
            for breakdown in breakdowns
        ]

    def calculate(self):
        with self.timings.measure("to_query"):
            query = self.to_query()

        with self.timings.measure("execute_hogql_query"):
            response = execute_hogql_query(
                query_type="revenue_analytics_arpu_query",
                query=query,
                team=self.team,
                timings=self.timings,
                modifiers=self.modifiers,
                limit_context=self.limit_context,
            )

        with self.timings.measure("build_results"):
            results = self._build_results(response)

        return RevenueAnalyticsArpuQueryResponse(
            results=results,
            hogql=response.hogql,
            modifiers=self.modifiers,
        )
