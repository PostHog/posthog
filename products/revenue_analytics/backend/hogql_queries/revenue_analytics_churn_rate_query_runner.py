from posthog.hogql import ast
from posthog.hogql.query import execute_hogql_query
from posthog.schema import (
    CachedRevenueAnalyticsChurnRateQueryResponse,
    RevenueAnalyticsChurnRateQueryResponse,
    RevenueAnalyticsChurnRateQuery,
)

from .revenue_analytics_query_runner import RevenueAnalyticsQueryRunner


class RevenueAnalyticsChurnRateQueryRunner(RevenueAnalyticsQueryRunner):
    query: RevenueAnalyticsChurnRateQuery
    response: RevenueAnalyticsChurnRateQueryResponse
    cached_response: CachedRevenueAnalyticsChurnRateQueryResponse

    def to_query(self) -> ast.SelectQuery:
        return ast.SelectQuery.empty()

    def calculate(self):
        response = execute_hogql_query(
            query_type="revenue_analytics_churn_rate_query",
            query=self.to_query(),
            team=self.team,
            timings=self.timings,
            modifiers=self.modifiers,
            limit_context=self.limit_context,
        )

        assert response.results

        return RevenueAnalyticsChurnRateQueryResponse(
            results=response.results,
            modifiers=self.modifiers,
        )
