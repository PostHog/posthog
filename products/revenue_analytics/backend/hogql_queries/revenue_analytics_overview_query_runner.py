from posthog.hogql import ast
from posthog.hogql.query import execute_hogql_query
from posthog.schema import (
    RevenueAnalyticsOverviewItem,
    RevenueAnalyticsOverviewItemKey,
    CachedRevenueAnalyticsOverviewQueryResponse,
    RevenueAnalyticsOverviewQueryResponse,
    RevenueAnalyticsOverviewQuery,
)

from .revenue_analytics_query_runner import RevenueAnalyticsQueryRunner
from posthog.hogql.database.schema.exchange_rate import EXCHANGE_RATE_DECIMAL_PRECISION

CONSTANT_ZERO = ast.Constant(value=0)


class RevenueAnalyticsOverviewQueryRunner(RevenueAnalyticsQueryRunner):
    query: RevenueAnalyticsOverviewQuery
    response: RevenueAnalyticsOverviewQueryResponse
    cached_response: CachedRevenueAnalyticsOverviewQueryResponse

    def to_query(self) -> ast.SelectQuery:
        # If there are no charge revenue views, we return a query that returns 0 for all values
        charge_subquery, _ = self.revenue_subqueries()
        if charge_subquery is None:
            return ast.SelectQuery(
                select=[
                    ast.Alias(alias="revenue", expr=CONSTANT_ZERO),
                    ast.Alias(alias="paying_customer_count", expr=CONSTANT_ZERO),
                    ast.Alias(alias="avg_revenue_per_customer", expr=CONSTANT_ZERO),
                ],
            )

        return ast.SelectQuery(
            select=[
                ast.Alias(
                    alias="revenue",
                    expr=ast.Call(
                        name="toDecimal",
                        args=[
                            ast.Call(name="sum", args=[ast.Field(chain=["amount"])]),
                            ast.Constant(value=EXCHANGE_RATE_DECIMAL_PRECISION),
                        ],
                    ),
                ),
                ast.Alias(alias="paying_customer_count", expr=ast.Call(name="count", args=[ast.Field(chain=["id"])])),
                ast.Alias(
                    alias="avg_revenue_per_customer",
                    expr=ast.Call(
                        name="ifNull",
                        args=[
                            ast.Call(
                                name="divideDecimal",
                                args=[
                                    ast.Field(chain=["revenue"]),
                                    ast.Call(
                                        name="toDecimal",
                                        args=[
                                            ast.Field(chain=["paying_customer_count"]),
                                            ast.Constant(value=EXCHANGE_RATE_DECIMAL_PRECISION),
                                        ],
                                    ),
                                ],
                            ),
                            CONSTANT_ZERO,
                        ],
                    ),
                ),
            ],
            select_from=ast.JoinExpr(table=charge_subquery),
            where=self.timestamp_where_clause(),
        )

    def calculate(self):
        response = execute_hogql_query(
            query_type="revenue_analytics_overview_query",
            query=self.to_query(),
            team=self.team,
            timings=self.timings,
            modifiers=self.modifiers,
            limit_context=self.limit_context,
        )

        assert response.results

        results = map_to_results(response.results)

        return RevenueAnalyticsOverviewQueryResponse(
            results=results,
            modifiers=self.modifiers,
        )


def map_to_results(results: list[dict]) -> list[RevenueAnalyticsOverviewItem]:
    result = results[0]  # Only care about the first result
    return [
        RevenueAnalyticsOverviewItem(key=RevenueAnalyticsOverviewItemKey.REVENUE, value=result[0]),
        RevenueAnalyticsOverviewItem(key=RevenueAnalyticsOverviewItemKey.PAYING_CUSTOMER_COUNT, value=result[1]),
        RevenueAnalyticsOverviewItem(key=RevenueAnalyticsOverviewItemKey.AVG_REVENUE_PER_CUSTOMER, value=result[2]),
    ]
