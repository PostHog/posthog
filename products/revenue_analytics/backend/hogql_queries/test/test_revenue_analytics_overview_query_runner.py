from freezegun import freeze_time
from pathlib import Path
from decimal import Decimal

from posthog.models.utils import uuid7
from products.revenue_analytics.backend.hogql_queries.revenue_analytics_overview_query_runner import (
    RevenueAnalyticsOverviewQueryRunner,
)
from products.revenue_analytics.backend.models import STRIPE_DATA_WAREHOUSE_CHARGE_IDENTIFIER
from posthog.schema import (
    CurrencyCode,
    DateRange,
    RevenueSources,
    RevenueAnalyticsOverviewQuery,
    RevenueAnalyticsOverviewQueryResponse,
    RevenueAnalyticsOverviewItemKey,
    RevenueAnalyticsOverviewItem,
)
from posthog.test.base import (
    APIBaseTest,
    ClickhouseTestMixin,
    _create_event,
    _create_person,
    snapshot_clickhouse_queries,
)
from posthog.warehouse.models import ExternalDataSchema

from posthog.warehouse.test.utils import create_data_warehouse_table_from_csv
from products.revenue_analytics.backend.hogql_queries.test.data.structure import (
    REVENUE_ANALYTICS_CONFIG_SAMPLE_EVENT,
    STRIPE_CHARGE_COLUMNS,
)

TEST_BUCKET = "test_storage_bucket-posthog.revenue_analytics.overview_query_runner.stripe_charges"


@snapshot_clickhouse_queries
class TestRevenueAnalyticsOverviewQueryRunner(ClickhouseTestMixin, APIBaseTest):
    QUERY_TIMESTAMP = "2025-02-15"

    def _create_purchase_events(self, data):
        person_result = []
        for distinct_id, timestamps in data:
            with freeze_time(timestamps[0][0]):
                person = _create_person(
                    team_id=self.team.pk,
                    distinct_ids=[distinct_id],
                    properties={
                        "name": distinct_id,
                        **({"email": "test@posthog.com"} if distinct_id == "test" else {}),
                    },
                )
            event_ids: list[str] = []
            for timestamp, session_id, revenue, currency in timestamps:
                event_ids.append(
                    _create_event(
                        team=self.team,
                        event="purchase",
                        distinct_id=distinct_id,
                        timestamp=timestamp,
                        properties={
                            "$session_id": session_id,
                            "revenue": revenue,
                            "currency": currency,
                        },
                    )
                )
            person_result.append((person, event_ids))
        return person_result

    def setUp(self):
        super().setUp()

        self.csv_path = Path(__file__).parent / "data" / "stripe_charges.csv"
        self.table, self.source, self.credential, self.csv_df, self.cleanUpFilesystem = (
            create_data_warehouse_table_from_csv(
                self.csv_path,
                "stripe_charge",
                STRIPE_CHARGE_COLUMNS,
                TEST_BUCKET,
                self.team,
            )
        )

        # Besides the default creations above, also create the external data schema
        # because this is required by the `RevenueAnalyticsRevenueView` to find the right tables
        self.schema = ExternalDataSchema.objects.create(
            team=self.team,
            name=STRIPE_DATA_WAREHOUSE_CHARGE_IDENTIFIER,
            source=self.source,
            table=self.table,
            should_sync=True,
            last_synced_at="2024-01-01",
        )

        self.team.revenue_analytics_config.base_currency = CurrencyCode.GBP.value
        self.team.revenue_analytics_config.events = [REVENUE_ANALYTICS_CONFIG_SAMPLE_EVENT]
        self.team.revenue_analytics_config.save()

    def tearDown(self):
        self.cleanUpFilesystem()
        super().tearDown()

    def _run_revenue_analytics_overview_query(
        self,
        date_range: DateRange | None = None,
        revenue_sources: RevenueSources | None = None,
    ):
        if date_range is None:
            date_range = DateRange(date_from="-30d")
        if revenue_sources is None:
            revenue_sources = RevenueSources(events=[], dataWarehouseSources=[str(self.source.id)])

        with freeze_time(self.QUERY_TIMESTAMP):
            query = RevenueAnalyticsOverviewQuery(dateRange=date_range, revenueSources=revenue_sources)
            runner = RevenueAnalyticsOverviewQueryRunner(
                team=self.team,
                query=query,
            )

            response = runner.calculate()
            RevenueAnalyticsOverviewQueryResponse.model_validate(response)

            return response

    def test_no_crash_when_no_data(self):
        self.table.delete()
        results = self._run_revenue_analytics_overview_query().results

        self.assertEqual(
            results,
            [
                RevenueAnalyticsOverviewItem(key=RevenueAnalyticsOverviewItemKey.REVENUE, value=0.0),
                RevenueAnalyticsOverviewItem(key=RevenueAnalyticsOverviewItemKey.PAYING_CUSTOMER_COUNT, value=0.0),
                RevenueAnalyticsOverviewItem(key=RevenueAnalyticsOverviewItemKey.AVG_REVENUE_PER_CUSTOMER, value=0.0),
            ],
        )

    def test_with_data(self):
        results = self._run_revenue_analytics_overview_query().results

        self.assertEqual(
            results,
            [
                RevenueAnalyticsOverviewItem(
                    key=RevenueAnalyticsOverviewItemKey.REVENUE, value=Decimal("1349.3495305777")
                ),
                RevenueAnalyticsOverviewItem(key=RevenueAnalyticsOverviewItemKey.PAYING_CUSTOMER_COUNT, value=5),
                RevenueAnalyticsOverviewItem(
                    key=RevenueAnalyticsOverviewItemKey.AVG_REVENUE_PER_CUSTOMER, value=Decimal("269.8699061155")
                ),
            ],
        )

    def test_with_events_data(self):
        s1 = str(uuid7("2023-12-02"))
        s2 = str(uuid7("2024-01-03"))
        s3 = str(uuid7("2024-02-04"))
        self._create_purchase_events(
            [
                ("p1", [("2023-12-02", s1, 4200, "USD")]),
                ("p2", [("2024-01-01", s2, 4300, "BRL"), ("2024-01-02", s3, 8700, "BRL")]),  # 2 events, 1 customer
            ]
        )

        results = self._run_revenue_analytics_overview_query(
            date_range=DateRange(date_from="2023-11-01", date_to="2024-01-31"),
            revenue_sources=RevenueSources(events=["purchase"], dataWarehouseSources=[]),
        ).results

        self.assertEqual(
            results,
            [
                RevenueAnalyticsOverviewItem(
                    key=RevenueAnalyticsOverviewItemKey.REVENUE, value=Decimal("54.2331251204")
                ),
                RevenueAnalyticsOverviewItem(key=RevenueAnalyticsOverviewItemKey.PAYING_CUSTOMER_COUNT, value=2),
                RevenueAnalyticsOverviewItem(
                    key=RevenueAnalyticsOverviewItemKey.AVG_REVENUE_PER_CUSTOMER, value=Decimal("27.1165625602")
                ),
            ],
        )
