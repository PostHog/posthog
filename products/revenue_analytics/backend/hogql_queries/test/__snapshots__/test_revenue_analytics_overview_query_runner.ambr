# serializer version: 1
# name: TestRevenueAnalyticsOverviewQueryRunner.test_no_crash_when_no_data
  '''
  SELECT 0 AS revenue,
         0 AS paying_customer_count,
         0 AS avg_revenue_per_customer
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295
  '''
# ---
# name: TestRevenueAnalyticsOverviewQueryRunner.test_with_data
  '''
  SELECT accurateCastOrNull(sum(amount), 'Decimal64(10)') AS revenue,
         count(DISTINCT customer_id) AS paying_customer_count,
         ifNull(divideDecimal(revenue, accurateCastOrNull(paying_customer_count, 'Decimal64(10)')), 0) AS avg_revenue_per_customer
  FROM
    (SELECT `stripe.posthog_test.charge_revenue_view`.id AS id,
            `stripe.posthog_test.charge_revenue_view`.timestamp AS timestamp,
            `stripe.posthog_test.charge_revenue_view`.customer_id AS customer_id,
            `stripe.posthog_test.charge_revenue_view`.session_id AS session_id,
            `stripe.posthog_test.charge_revenue_view`.event_name AS event_name,
            `stripe.posthog_test.charge_revenue_view`.original_currency AS original_currency,
            `stripe.posthog_test.charge_revenue_view`.original_amount AS original_amount,
            `stripe.posthog_test.charge_revenue_view`.currency_is_zero_decimal AS currency_is_zero_decimal,
            `stripe.posthog_test.charge_revenue_view`.amount_decimal_divider AS amount_decimal_divider,
            `stripe.posthog_test.charge_revenue_view`.adjusted_original_amount AS adjusted_original_amount,
            `stripe.posthog_test.charge_revenue_view`.currency AS currency,
            `stripe.posthog_test.charge_revenue_view`.amount AS amount
     FROM
       (SELECT posthog_test_stripe_charge.id AS id,
               parseDateTime64BestEffortOrNull(toString(posthog_test_stripe_charge.created), 6, 'UTC') AS timestamp,
               posthog_test_stripe_charge.customer AS customer_id,
               NULL AS session_id,
               NULL AS event_name,
               upper(posthog_test_stripe_charge.currency) AS original_currency,
               accurateCastOrNull(posthog_test_stripe_charge.amount_captured, 'Decimal64(10)') AS original_amount,
               in(original_currency,
                  ['BIF', 'CLP', 'DJF', 'GNF', 'JPY', 'KMF', 'KRW', 'MGA', 'PYG', 'RWF', 'UGX', 'VND', 'VUV', 'XAF', 'XOF', 'XPF']) AS currency_is_zero_decimal,
                 if(currency_is_zero_decimal, accurateCastOrNull(1, 'Decimal64(10)'), accurateCastOrNull(100, 'Decimal64(10)')) AS amount_decimal_divider,
                 divideDecimal(original_amount, amount_decimal_divider) AS adjusted_original_amount,
                 'GBP' AS currency,
                 if(equals(original_currency, currency), toDecimal64(adjusted_original_amount, 10), if(dictGetOrDefault(`posthog_test`.`exchange_rate_dict`, 'rate', original_currency, toDate(ifNull(timestamp, toDateTime(0, 'UTC'))), toDecimal64(0, 10)) = 0, toDecimal64(0, 10), multiplyDecimal(divideDecimal(toDecimal64(adjusted_original_amount, 10), dictGetOrDefault(`posthog_test`.`exchange_rate_dict`, 'rate', original_currency, toDate(ifNull(timestamp, toDateTime(0, 'UTC'))), toDecimal64(0, 10))), dictGetOrDefault(`posthog_test`.`exchange_rate_dict`, 'rate', currency, toDate(ifNull(timestamp, toDateTime(0, 'UTC'))), toDecimal64(0, 10))))) AS amount
        FROM s3('http://host.docker.internal:19000/posthog/test_storage_bucket-posthog.revenue_analytics.overview_query_runner.stripe_charges/posthog_test_stripe_charge/*.csv', 'object_storage_root_user', 'object_storage_root_password', 'CSVWithNames', '`id` String, `paid` Int8, `amount` Int64, `object` String, `source` String, `status` String, `created` DateTime, `invoice` String, `outcome` String, `captured` Int8, `currency` String, `customer` String, `disputed` Int8, `livemode` Int8, `metadata` String, `refunded` Int8, `description` String, `receipt_url` String, `failure_code` String, `fraud_details` String, `radar_options` String, `receipt_email` String, `payment_intent` String, `payment_method` String, `amount_captured` Int64, `amount_refunded` Int64, `billing_details` String, `failure_message` String, `balance_transaction` String, `statement_descriptor` String, `payment_method_details` String, `calculated_statement_descriptor` String') AS posthog_test_stripe_charge
        WHERE equals(posthog_test_stripe_charge.status, 'succeeded')) AS `stripe.posthog_test.charge_revenue_view`)
  WHERE and(ifNull(greaterOrEquals(timestamp, assumeNotNull(toDateTime('2025-01-16 00:00:00', 'UTC'))), 0), ifNull(lessOrEquals(timestamp, assumeNotNull(toDateTime('2025-02-15 23:59:59', 'UTC'))), 0))
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295
  '''
# ---
# name: TestRevenueAnalyticsOverviewQueryRunner.test_with_events_data
  '''
  SELECT accurateCastOrNull(sum(amount), 'Decimal64(10)') AS revenue,
         count(DISTINCT customer_id) AS paying_customer_count,
         ifNull(divideDecimal(revenue, accurateCastOrNull(paying_customer_count, 'Decimal64(10)')), 0) AS avg_revenue_per_customer
  FROM
    (SELECT events_revenue_view.id AS id,
            events_revenue_view.timestamp AS timestamp,
            events_revenue_view.customer_id AS customer_id,
            events_revenue_view.session_id AS session_id,
            events_revenue_view.event_name AS event_name,
            events_revenue_view.original_currency AS original_currency,
            events_revenue_view.original_amount AS original_amount,
            events_revenue_view.currency_is_zero_decimal AS currency_is_zero_decimal,
            events_revenue_view.amount_decimal_divider AS amount_decimal_divider,
            events_revenue_view.adjusted_original_amount AS adjusted_original_amount,
            events_revenue_view.currency AS currency,
            events_revenue_view.amount AS amount
     FROM
       (SELECT events.uuid AS id,
               toTimeZone(events.created_at, 'UTC') AS timestamp,
               events.distinct_id AS customer_id,
               nullIf(nullIf(events.`$session_id`, ''), 'null') AS session_id,
               events.event AS event_name,
               multiIf(equals(events.event, 'purchase'), upper(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(events.properties, 'currency'), ''), 'null'), '^"|"$', '')), NULL) AS original_currency,
               multiIf(equals(events.event, 'purchase'), accurateCastOrNull(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(events.properties, 'revenue'), ''), 'null'), '^"|"$', ''), 'Decimal64(10)'), NULL) AS original_amount,
               in(original_currency,
                  ['BIF', 'CLP', 'DJF', 'GNF', 'JPY', 'KMF', 'KRW', 'MGA', 'PYG', 'RWF', 'UGX', 'VND', 'VUV', 'XAF', 'XOF', 'XPF']) AS currency_is_zero_decimal,
                 if(currency_is_zero_decimal, accurateCastOrNull(1, 'Decimal64(10)'), accurateCastOrNull(100, 'Decimal64(10)')) AS amount_decimal_divider,
                 divideDecimal(original_amount, amount_decimal_divider) AS adjusted_original_amount,
                 'GBP' AS currency,
                 multiIf(equals(events.event, 'purchase'), if(isNull(upper(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(events.properties, 'currency'), ''), 'null'), '^"|"$', ''))), accurateCastOrNull(adjusted_original_amount, 'Decimal64(10)'), if(equals(upper(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(events.properties, 'currency'), ''), 'null'), '^"|"$', '')), 'GBP'), toDecimal64(adjusted_original_amount, 10), if(dictGetOrDefault(`posthog_test`.`exchange_rate_dict`, 'rate', upper(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(events.properties, 'currency'), ''), 'null'), '^"|"$', '')), toDate(toTimeZone(events.timestamp, 'UTC')), toDecimal64(0, 10)) = 0, toDecimal64(0, 10), multiplyDecimal(divideDecimal(toDecimal64(adjusted_original_amount, 10), dictGetOrDefault(`posthog_test`.`exchange_rate_dict`, 'rate', upper(replaceRegexpAll(nullIf(nullIf(JSONExtractRaw(events.properties, 'currency'), ''), 'null'), '^"|"$', '')), toDate(toTimeZone(events.timestamp, 'UTC')), toDecimal64(0, 10))), dictGetOrDefault(`posthog_test`.`exchange_rate_dict`, 'rate', 'GBP', toDate(toTimeZone(events.timestamp, 'UTC')), toDecimal64(0, 10)))))), NULL) AS amount
        FROM events
        WHERE and(equals(events.team_id, 99999), and(equals(events.event, 'purchase'), isNotNull(amount)))
        ORDER BY timestamp DESC) AS events_revenue_view
     WHERE in(events_revenue_view.event_name,
              ['purchase']))
  WHERE and(ifNull(greaterOrEquals(timestamp, assumeNotNull(toDateTime('2023-11-01 00:00:00', 'UTC'))), 0), ifNull(lessOrEquals(timestamp, assumeNotNull(toDateTime('2024-01-31 23:59:59', 'UTC'))), 0))
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295
  '''
# ---
